File: Notes/Structures.cpp
// segment tree with point updates

struct SegTree{
	vector<ll> tree;
	int n;
	
	SegTree(vi &arr,ll sz){
		n = sz;
		tree.resize(2*n);
		loop(i,0,n-1,1) tree[i+n] = arr[i];
		loopR(i,n-1,1,1) tree[i] = tree[2*i] + tree[2*i + 1];
	}

	void update(int idx, ll val){
		idx--;
		idx+=n;
		tree[idx] = val;
		idx>>=1;
		while(idx){
			tree[idx] = tree[2*idx] + tree[2*idx + 1];
			idx>>=1;
		}
	}

	ll query(int l, int r){ // range [l,r)
		l--;
		r--;
		l+=n;
		r+=n;
		ll sum = 0;
		while (l < r){
			if (l&1) sum+=tree[l++];
			if (r&1) sum+=tree[--r];
			l>>=1;
			r>>=1;
		}
		return sum;
	}
};

// DSU

struct DSU{
	vector<int> parent,low,high;
	ll comps;

	DSU(int n){
		parent.resize(n+1);
		iota(all(parent),0);
		low=high=parent;
		comps=n;
	}

	int FIND(int x){
		if (parent[x]==x) return x;
		return parent[x]=FIND(parent[x]);
	}

	bool UNION(int u, int v){
		u=FIND(u),v=FIND(v);
		if (u==v) return false;
		low[u]=min(low[u],low[v]);
		high[u]=max(high[u],high[v]);
		parent[v]=u;
		comps--;
		return true;
	}

	int getHigh(int x){
		return high[FIND(x)];
	}

	int getLow(int x){
		return low[FIND(x)];
	}
};

// minQueue

struct minQ{
	deque<int> dq;
	
	void PUSH(int ind){
		while(dq.size() and arr[dq.back()] >= arr[ind]) dq.pop_back();
		dq.push_back(ind);
	}

	void POP(int ind){
		while(dq.size() and dq.front() <= ind) dq.pop_front();
	}

	int front(){
		return dq.front();
	}
};

// BIT

struct BIT{
	int n;
	vector<int> tree;

	BIT(int sz){
		n = sz;
		tree.resize(n+1);
	}
	void update(int idx,int val){
		while(idx <= n){
			tree[idx]+=val;
			idx += (idx & -idx);
		}
	}
	int query(int idx){
		int sum = 0;
		while(idx){
			sum += tree[idx];
			idx -= (idx & -idx);
		}
		return sum;
	}
	int sum(int l, int r){
		return query(r)-query(l-1);
	}
	int getKth(int k){
		int low=1,high=n,mid;
		while(low < high){
			mid=(low+high)/2;
			if(query(mid) >= k) high=mid;
			else low=mid+1;
		}
		return low;
	}
};

File: Notes/Python/Structures.py
# Segment Tree with point updates


class SegTree:

    def __init__(self, arr, n):
        self.n = n
        self.tree = [0 for x in range(n)] + [x for x in arr]
        for x in range(n-1, 0, -1):
            self.tree[x] = self.tree[2*x] + self.tree[2*x + 1]

    def query(self, l, r):    # sum of range [l,r)
        l -= 1
        r -= 1
        l += n
        r += n
        s = 0
        while l < r:
            if l & 1:
                s += self.tree[l]
                l += 1
            if r & 1:
                r -= 1
                s += self.tree[r]
            l = l >> 1
            r = r >> 1
        return s

    def update(self, idx, val):
        idx -= 1
        idx += self.n
        self.tree[idx] = val
        idx = idx >> 1
        while idx:
            self.tree[idx] = self.tree[2*idx] + self.tree[2*idx + 1]
            idx = idx >> 1

    def getKth(self, k):
        low, high = 1, self.n
        while low < high:
            mid = (low+high)//2
            if self.query(1, mid) < k:
                low = mid + 1
            else:
                high = mid
        return low

# BIT with point updates


class BIT:

    def __init__(self, arr, n):
        self.n = n
        self.tree = [0 for x in range(self.n+1)]

    def update(self, idx, val):
        while idx <= self.n:
            self.tree[idx] += val
            idx += (idx & -idx)

    def query(self, idx):
        s = 0
        while idx:
            s += self.tree[idx]
            idx -= (idx & -idx)
        return s

    def getKth(self, k):
        low, high = 1, self.n
        while low < high:
            mid = (low+high)//2
            if self.query(mid) < k:
                low = mid + 1
            else:
                high = mid
        return low

# DSU with size and path compression heuristics


class DSU:

    def __init__(self, n):
        self.n = n
        self.comps = n
        self.parent = [i for i in range(n)]
        self.size = [1 for i in range(n)]

    def FIND(self, x):
        if x != self.parent[x]:
            self.parent[x] = self.FIND(self.parent[x])
        return self.parent[x]

    def UNION(self, x, y):
        x, y = self.FIND(x), self.FIND(y)
        if x == y:
            return 0
        self.comps -= 1
        if self.size[x] > self.size[y]:
            x, y = y, x
        self.parent[x] = y
        self.size[y] += self.size[x]
        return 1


class Trie:

    def __init__(self):
        self.arr = {}
        self.count = 0
        self.end = 0

    def insert(self, word):
        for x in word:
            if x not in self.arr:
                self.arr[x] = Trie()
            self = self.arr[x]
            self.count += 1
        self.end = True

    def findPrefix(self, word):
        for x in word:
            if x not in self.arr:
                return 0, 0
            self = self.arr[x]
        return self.end, self.count


class BitTrie:

    def __init__(self):
        self.arr = [None, None]

    def insert(self, num):
        for x in range(31, -1, -1):
            bit = (num >> x) & 1
            if self.arr[bit] is None:
                self.arr[bit] = BitTrie()
            self = self.arr[bit]
        self.ends = True

    def XorMaxQuery(self, num):
        ans = 0
        for x in range(31, -1, -1):
            bit = (num >> x) & 1
            bit = bit ^ 1 if self.arr[bit ^ 1] else bit
            self = self.arr[bit]
            ans = (ans << 1) | bit
        return ans ^ num
